### PR TITLE
Add test for contract

### PR DESCRIPTION
**Pull request by GNaD13**
#660 

add test case: Can a contract close a wasm ibc channel that is not owned by the contract?

### CODE REVIEW

File: x/wasm/keeper/keeper.go [hunk 1]
```diff
@@ -960,6 +960,26 @@ func (k Keeper) setAccessConfig(ctx sdk.Context, codeID uint64, caller sdk.AccAd
 	return nil
 }
 
+func (k *Keeper) HandleContractResponse(
```
Comment (jhernandezb) [inline x/wasm/keeper/keeper.go:963]: I don't think we should expose this, or at least expose it for tests only.

Comment (GNaD13) [reply x/wasm/keeper/keeper.go:963]: tysm, I'll fix it

Comment (GNaD13) [reply x/wasm/keeper/keeper.go]: Hi @jhernandezb , I have fixed this problem. Tysm for your reviews. I've learned a lot <3

**Review by ethanfrey — COMMENTED** [review]
Looks like a useful test case. Thank you for the contribution.

I would wait for @alpe to review it for code quality before merge, but he is off sick this week.

**Comment by alpe** [comment]
I will 👁️ at this tomorrow when I feel better. 🦷 OP.

File: x/wasm/relay_test.go [hunk 2]
```diff
@@ -497,6 +557,43 @@ func (c *sendEmulatedIBCTransferContract) IBCPacketTimeout(codeID wasmvm.Checksu
 	return &wasmvmtypes.IBCBasicResponse{Messages: []wasmvmtypes.SubMsg{{ReplyOn: wasmvmtypes.ReplyNever, Msg: wasmvmtypes.CosmosMsg{Bank: returnTokens}}}}, 0, nil
 }
 
+var _ wasmtesting.IBCContractCallbacks = &captureCloseContractNotOwned{}
+
+type captureCloseContractNotOwned struct {
```
Comment (alpe) [inline x/wasm/relay_test.go:562]: Nit: it is not capturing anything. How about:
```suggestion
type closeChannelContract struct {
```

File: x/wasm/relay_test.go [hunk 3]
```diff
@@ -497,6 +557,43 @@ func (c *sendEmulatedIBCTransferContract) IBCPacketTimeout(codeID wasmvm.Checksu
...
+	if err := json.Unmarshal(executeMsg, &in); err != nil {
+		return nil, 0, err
+	}
+	ibcMsg := &wasmvmtypes.IBCMsg{
+		CloseChannel: &wasmvmtypes.CloseChannelMsg{
+			ChannelID: in.ChannelID,
+		},
+	}
+
+	return &wasmvmtypes.Response{Messages: []wasmvmtypes.SubMsg{{ReplyOn: wasmvmtypes.ReplyNever, Msg: wasmvmtypes.CosmosMsg{IBC: ibcMsg}}}}, 0, nil
```
Comment (alpe) [inline x/wasm/relay_test.go:582]: very nice!

File: x/wasm/relay_test.go [hunk 4]
```diff
@@ -497,6 +557,43 @@ func (c *sendEmulatedIBCTransferContract) IBCPacketTimeout(codeID wasmvm.Checksu
...
+}
+
+type closeChannelIbc struct {
```
Comment (alpe) [inline x/wasm/relay_test.go:585]: nit: uppercase IBC
```suggestion
type closeIBCChannel struct {
```

File: x/wasm/relay_test.go [hunk 5]
```diff
@@ -399,6 +399,66 @@ func TestContractHandlesChannelClose(t *testing.T) {
...
+	}
+	path.EndpointB.ChannelConfig = &ibctesting.ChannelConfig{
+		PortID:  chainB.ContractInfo(myContractAddrB).IBCPortID,
+		Version: ibctransfertypes.Version,
+		Order:   channeltypes.UNORDERED,
+	}
+	coordinator.SetupConnections(path)
+	coordinator.CreateChannels(path)
+
+	closeChannelIbcMsg := &types.MsgExecuteContract{
```
Comment (alpe) [inline x/wasm/relay_test.go:449]: Good setup!

File: x/wasm/ibctesting/chain.go [hunk 6]
```diff
@@ -273,6 +273,41 @@ func (chain *TestChain) SendMsgs(msgs ...sdk.Msg) (*sdk.Result, error) {
...
+	_, r, err := app.SignAndDeliver(
+		chain.t,
+		chain.TxConfig,
+		chain.App.BaseApp,
+		chain.GetContext().BlockHeader(),
+		msgs,
+		chain.ChainID,
+		[]uint64{chain.SenderAccount.GetAccountNumber()},
+		[]uint64{chain.SenderAccount.GetSequence()},
+		true, expPass, chain.SenderPrivKey,
```
Comment (alpe) [inline x/wasm/ibctesting/chain.go:289]: SignAndDeliver is a monster :-)

**Review by alpe — APPROVED** [review]
Thanks for providing this test case. 🌷 
The IBC mock tests are super straight forward but you did this very well! I have marked some minor nits, if you want to polish your code. But otherwise this is great! Happy to merge it.

File: x/wasm/relay_test.go (hunk 2 above)
Comment (GNaD13) [reply x/wasm/relay_test.go]: I've changed this

File: x/wasm/relay_test.go (hunk 3 above)
Comment (GNaD13) [reply x/wasm/relay_test.go:582]: Thank you!!

File: x/wasm/relay_test.go (hunk 4 above)
Comment (GNaD13) [reply x/wasm/relay_test.go]: I've changed this

**Comment by GNaD13** [comment]
> Thanks for providing this test case. tulip The IBC mock tests are not super straight forward but you did this very well! I have marked some minor nits, if you want to polish your code. But otherwise this is great! Happy to merge it.

tysm for your review :3

**Pull request merged**